### PR TITLE
When using PinPad, show timeout warning to user

### PR DESCRIPTION
**Pull request by olegprokofjev**
IB-4807:
   When user is using PinPad for either loading picture or changing PIN1
   or PIN2 or PUK, show the warning message on timeout (also if user
   presses Cancel button).
   It will work the similar way as in qdigidoc (which is using as card
   driver the MS Cryptography API).

Signed-off-by: Oleg Prokofjev <oleg@aktors.ee>

### CODE REVIEW

**Comment by metsma** [comment]
Why not show progress bar like DigiDoc does?

**Comment by olegprokofjev** [comment]
qesteidutil.exe (seems tera.exe as well) is using Common\QPCSC.cpp driver.
Partly it shows the own made progress bar (Load Picture, @eesti.ee email, Mobile-ID, Update certificate, ...), partly it does not (Changing PIN/PUK codes). I do not know the background - why it was done so.

The qdigidocclient.exe is using MS functionality (like NCryptSignHash) which shows the progress bar on language, defined in computer's Region & Language settings (Win).

**Comment by metsma** [comment]
But DigiDoc on OSX/Linux shows custom Progress Bar

**Comment by olegprokofjev** [comment]
Yes, for OSX/Linux the qDigiDoc is using different functionality.
I can try to show the progress bar on change of PIN/PUK codes in qEstEidUtil although it was not requested in ticket.

File: src/MainWindow.cpp [hunk 1]
```diff
@@ -196,7 +196,17 @@ bool MainWindowPrivate::validateCardError( QSmartCardData::PinType type, int fla
...
+		if ( q->d->smartcard->data().isPinpad() )
+		{
+			switch ( type )
+			{
+			case QSmartCardData::Pin1Type: showWarning( tr("PIN1 timeout") ); break;
+			case QSmartCardData::Pin2Type: showWarning( tr("PIN2 timeout") ); break;
+			case QSmartCardData::PukType: showWarning( tr("PUK timeout") ); break;
+			}
+		}
+		break;
```
Comment (metsma) [inline src/MainWindow.cpp]: why additional dialog? when progress bar disappears, it should already indicate that its expired

File: src/QSmartCard.cpp [hunk 2]
```diff
@@ -233,20 +233,33 @@ QSmartCard::~QSmartCard()
 	delete d;
 }
 
-QSmartCard::ErrorType QSmartCard::change(QSmartCardData::PinType type, const QString &newpin, const QString &pin)
+QSmartCard::ErrorType QSmartCard::change(QSmartCardData::PinType type, const QString &newpin, const QString &pin, const QString &title, const QString &bodyText)
 {
 	QMutexLocker locker(&d->m);
-	QSharedPointer<QPCSCReader> reader(d->connect(d->t.reader()));
+ 	QSharedPointer<QPCSCReader> reader(d->connect(d->t.reader()));
```
Comment (metsma) [inline src/QSmartCard.cpp]: Whitespace change?

File: src/MainWindow.cpp (hunk 1 above)
Comment (olegprokofjev) [inline src/MainWindow.cpp]: It was a main requirement of IB-4807. Same way it works in qDigidoc3 client. I suppose that both applications should work the same way. Also user can cancel operation (progress bar), but it continue to work in background until time limit and then the timeout message is shown, which is OK to my mind. Tells user that she/he can continue using application.

Comment (metsma) [inline src/MainWindow.cpp]: DigiDoc3 client does not show message on OSX/Linux

Comment (olegprokofjev) [inline src/MainWindow.cpp]: Correct. I will remove 'timeout' message dialog for OSX/Linux.

Comment (metsma) [inline src/MainWindow.cpp]: Same is also on windows, non estonian cards

**Comment by metsma** [comment]
qesteidutil is now feature complete and no new development is made.
[DigiDoc4-Client](/open-eid/DigiDoc4-Client) supersedes qdigidoc

**Pull request closed without merging**